scene/resources/dynamic_font.cpp

```diff
@@ -303,14 +326,142 @@ void DynamicFontAtSize::set_texture_flags(uint32_t p_flags) {
 	}
 }
 
+bool DynamicFontAtSize::shape_run(Vector<Ref<ClusterInfo> > &p_visual_order, Vector<Ref<ClusterInfo> > &p_logical_order, const Vector<UChar> &p_utf16_text, int p_start, int p_length, TextDirection p_direction, const Vector<hb_feature_t> &p_features, const String &p_script, const String &p_lang) const {
+
+	hb_buffer_t *buffer = hb_buffer_create();
+	hb_buffer_set_direction(buffer, (p_direction == DIR_LTR) ? HB_DIRECTION_LTR : HB_DIRECTION_RTL);
+	if (p_script.length() > 0) {
+		hb_buffer_set_script(buffer, hb_script_from_string(p_script.ascii().get_data(), -1));
+	} else {
+		//auto detect script
```

Comment:
This is rather simplistic. ICU has a bit more sophisticated code, but is not part of the public API AFAIK, you can find it in `source/extra/scrptrun/` in ICU source tree (I usually just copy the couple files and use them).

Script itemization should also be done on the whole text (just like bidi is done) not on individual runs, since bidi can result in separate runs of common characters only. Also I don’t think there is much value in having to set script explicitly in client code, it should always be auto-detected just like bidi as it can always be determined from character properties, unlike language.

Reply:
I won't be around till 18 sep., will look at it as soon as I'm back.